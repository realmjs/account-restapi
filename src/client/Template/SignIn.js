"use strict"

import React, { Component, PureComponent } from 'react'

import xhttp from '@realmjs/xhttp-request'

import Navigator from './CommonWidget/Navigator'
import BackButton from './CommonWidget/BackButton'

import { isEmail } from '../../lib/form'

/* Widgets used in Sign-in page */

class Email extends PureComponent {
  constructor(props) {
    super(props)
    this.state = { email: '', error: '', syncing: false }
    this.getTypedInput = this.getTypedInput.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.onConfirm = this.onConfirm.bind(this)
  }
  render() {
    return (
      <div>
        <header>
          <span onClick={this.props.close} className="w3-button w3-right w3-red" style={{ display: (window.self === window.top)? 'none': 'block'}}>&times;</span>
          <h3 className="w3-text-blue" style={{fontWeight: "bold"}} > Login </h3>
        </header>
        <p className="w3-text-blue-grey">
          Please enter your email
        </p>
        <p>
          <label className="w3-text-grey"> Email </label>
          <label className="w3-right w3-text-red"> {this.state.error || ''} </label>
          <input className = {`w3-input w3-border ${this.state.error && this.state.error.length > 0 ? 'w3-border-red' : ''}`}
                type = "text"
                placeholder = "email@example.com"
                value = {this.state.email}
                onChange = {this.getTypedInput}
                onKeyUp = {this.handleKeyUp}
                disabled = {this.state.syncing}
          />
        </p>
        <div style = {{marginBottom: '42px'}}>
          <div className="w3-cell-row">
            <div className="w3-cell">
              <label className="w3-text-red w3-large"><a href={`/form?name=signup&app=${__data.app}`}> Create New Account </a></label>
            </div>
            <div className="w3-cell" style={{textAlign: 'right'}}>
              <button type="submit" className={`w3-button w3-blue `} onClick={this.onConfirm} disabled = {this.state.syncing} >
                Next {' '}
                {
                  this.state.syncing? <i className ="fa fa-circle-o-notch fa-spin" /> : <i className ="fa fa-chevron-right" />
                }
              </button>
            </div>
          </div>
        </div>
      </div>
    )
  }
  getTypedInput(e) {
    this.setState({ email: e.target.value, error: '' })
  }
  handleKeyUp(e) {
    if (e.which === 13 || e.keyCode === 13) {
      this.onConfirm()
    }
  }
  onConfirm() {
    const email = this.state.email
    if (email.length === 0) {
      const error = "Email is empty"
      this.setState({ error })
      return
    }
    if (!isEmail(email)) {
      const error = "Invalid email"
      this.setState({ error })
      return
    }
    this.setState({ syncing: true })
    this.props.onConfirm('email')(email, error => this.setState({ error, syncing: false }))
  }
}

class Password extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      password: '',
      error: null,
      syncing: false
    }
    this.getTypedPassword = this.getTypedPassword.bind(this)
    this.handleKeyUpForPassword = this.handleKeyUpForPassword.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
  }
  render() {
    const form = this.props.form
    return (
      <div>
          <header >
            <span onClick={this.props.close} className="w3-button w3-right w3-red" style={{ display: (window.self === window.top)? 'none': 'block'}}>&times;</span>
            <BackButton onClick = {this.props.back} />
          </header>
          <div>
            <div className ="w3-text-blue" >
              <h3> {form.email} </h3>
            </div>
            <p>
              <label className="w3-text-grey">Password</label>
              <label className="w3-right w3-text-red"> {this.state.error || ''} </label>
              <input  className = {`w3-input w3-border ${this.state.error && this.state.error.length > 0 ? 'w3-border-red' : ''}`}
                      type = "password"
                      placeholder = "password"
                      value = {this.state.password}
                      onChange = {this.getTypedPassword}
                      onKeyUp = {this.handleKeyUpForPassword}
              />
            </p>
            <div style = {{marginBottom: '42px'}}>
              <div className="w3-cell-row">
                <div className="w3-cell">
                <label className="w3-text-orange "><a href={`/form?name=requestresetpassword&app=${__data.app}&email=${form.email}`}> Forgot your password </a></label>
                </div>
                <div className="w3-cell" style={{textAlign: 'right'}}>
                  <button className = {`w3-button w3-blue`}
                        onClick = {this.onSubmit} disabled = {this.state.syncing} >
                    Submit {' '}
                    {
                      this.state.syncing ?
                        <i className ="fa fa-circle-o-notch w3-spin" style = {{marginLeft: '4px'}} />
                      :
                      <i className ="fa fa-level-down fa-rotate-90" style = {{marginLeft: '4px'}} />
                    }
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>

    )
  }
  getTypedPassword(e) {
    this.setState({ password: e.target.value, error: '' })
  }

  handleKeyUpForPassword(e) {
    if (e.which == 13 || e.keyCode == 13) {
      this.onSubmit()
    }
  }
  onSubmit() {
    const password = this.state.password
    if (password.length === 0) {
      this.setState({ error: 'Password must not be empty'})
      return
    }
    this.setState({ syncing: true })
    this.props.onConfirm('password')(password, error => this.setState({ error, syncing: false }))
  }
}

class Welcome extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {}
  }
  static getDerivedStateFromProps(props, state) {
    if (props.active) {
      setTimeout( _ => props.close(), 1500)
      return null
    } else {
      return null
    }
  }
  render() {
    const user = this.props.user
    return (
      <div style = {{textAlign: 'center'}} >
        <div >
          <h3> Welcome {user && user.profile && user.profile.displayName} </h3>
        </div>
        <p>
          You have signed in.
        </p>
      </div>
    )
  }
}

/* Sign-In Page */

const routes = [
  { name: 'email', template: Email },
  { name: 'password', template: Password },
  { name: 'welcome', template: Welcome, animate: 'w3-animate-top' },
]

export default class SignIn extends Component {
  constructor(props) {
    super(props)
    this.state = {
      activeRoute: 'email',
      form: { email: '', password: '' },
      user: null
    }
    this.flow = ['email', 'password', 'welcome']
    const methods = ['navigate', 'next', 'back', 'onConfirm']
    methods.forEach( method => this[method] = this[method].bind(this) )
  }
  render() {
    return (
      <div className="w3-container" style={{ padding: "10px 12px", maxWidth: "460px" }}>
        <Navigator  routes = {routes}
                    activeRoute = {this.state.activeRoute}
                    navigate = {this.navigate}
                    next = {this.next}
                    back = {this.back}
                    onConfirm = {this.onConfirm}
                    form = {this.state.form}
                    {...this.props}

        />
      </div>
    )
  }
  navigate(route) {
    this.setState({ activeRoute: route })
    return this
  }
  next() {
    const current = this.flow.findIndex( route => route === this.state.activeRoute )
    if (current < this.flow.length- 1) {
      this.setState({ activeRoute: this.flow[current + 1] })
    }
    return this
  }
  back() {
    const current = this.flow.findIndex( route => route === this.state.activeRoute )
    if (current > 0) {
      this.setState({ activeRoute: this.flow[current - 1] })
    }
    return this
  }
  onConfirm(route) {
    return this[`onConfirm${route.replace(/^\w/, c => c.toUpperCase())}`].bind(this)
  }
  onConfirmEmail(email, done) {
    if (email === this.state.form.email) {
      done && done()
      this.next()
    }
    xhttp.get(`/users?u=${email}`, { timeout: 30000 })
    .then( ({status}) => {
      if (status === 200) {
        done && done()
        const form = {...this.state.form}
        form.email = email
        this.setState({ form })
        this.next()
      } else if (status === 404) {
        done && done(`Not registered`)
      } else {
        done && done(`Error: ${status}`)
        console.error(`Error returned by server: ${status}`)
      }
    })
    .catch( err => {
      done && done(`Error: Network timeout`)
    })
  }
  onConfirmPassword(password, done) {
    const form = {...this.state.form}
    xhttp.post(`/session`, { username: form.email, password }, { timeout: 30000 })
    .then( ({status, responseText}) => {
      if (status === 200) {
        done && done()
        const session = JSON.parse(responseText)
        this.setState({ user: session.user})
        this.props.done && this.props.done({status: 200, session})
        this.next()
      } else {
        done && done(`Error: Failed to signing in`)
      }
    })
    .catch( err => {
      done && done(`Error: Network timeout`)
    })
  }
}
