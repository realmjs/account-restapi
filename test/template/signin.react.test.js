"use strict"
import "core-js/stable";
import "regenerator-runtime/runtime";

import React from 'react';
import renderer from 'react-test-renderer';
import {screen, fireEvent, render} from '@testing-library/react';

import xhttp from '@realmjs/xhttp-request';
jest.mock('@realmjs/xhttp-request');

import SignIn from '../../src/client/template/SignIn';

const done = jest.fn(msg => msg);
const close = jest.fn(msg => msg);
const data = {"route":"signin","targetOrigin":"localhost:3100","app":"account","query":{"app":"account","name":"signin"}};

beforeEach( () => jest.clearAllMocks() );


function waitfor(cb) {
  return new Promise(resolve => {
    setTimeout(() => { cb(); resolve(); }, 0);
  });
}

test('Should match snapshot of first scene (email) after mounted', () => {
  const component = renderer.create(
    <SignIn data = {data} done = {done} close = {close} />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


test('Should alert invalid if entering wrong email format', () => {
  render(
    <SignIn data = {data} done = {done} close = {close} />
  );
  // attemp to enter invalid email format

  const inputEmailNode = screen.getByLabelText('email');
  testInValidEmail(inputEmailNode, 'wrongemail', 'Invalid email');
  testInValidEmail(inputEmailNode, 'email@', 'Invalid email');
  testInValidEmail(inputEmailNode, 'email@domain', 'Invalid email');
  testInValidEmail(inputEmailNode, '<script>alert("hacked")</script>', 'Invalid email');
  testInValidEmail(inputEmailNode, '', 'Email is empty');

  function testInValidEmail(node, text, error) {
    expect(screen.queryByText(error)).toBeNull();
    fireEvent.change(node, {target: { value: text } });
    fireEvent.keyUp(node, { keyCode: 13 });
    expect(screen.queryByText(error)).not.toBeNull();
    // when change value, error message should disapear
    fireEvent.change(node, {target: { value: 'a' } });
    expect(screen.queryByText(error)).toBeNull();
  }
});


test('Should alert Not registered if entering valid but unregistered email', async () => {

  xhttp.get.mockResolvedValue({ status: 404 });

  render(
    <SignIn data = {data} done = {done} close = {close} />
  );

  const inputEmailNode = screen.getByLabelText('email');
  await expectShowError(inputEmailNode, 'newtester@team.com', 'Not registered');
  await expectShowError(inputEmailNode, 'new.tester123@my.team.com', 'Not registered');

  function expectShowError(node, text, error) {
      fireEvent.change(node, {target: { value: '' } });
      expect(screen.queryByText(error)).toBeNull();
      fireEvent.change(node, {target: { value: text } });
      fireEvent.keyUp(node, { keyCode: 13 });
      expect(xhttp.get.mock.calls[0][0]).toMatch(/(\/user\?u=.*&app=.*|\/user\?app=.*&u=.*)/);
      return screen.findByText(error);
  }

});


test('Should move to Password scene after entering a registered email and back after clicking back button', async () => {

  xhttp.get.mockResolvedValue({ status: 200 });

  const { container } = render( <SignIn data = {data} done = {done} close = {close} /> );

  const inputEmailNode = screen.getByLabelText('email');
  const nextButtonNode = screen.getByText('Next');
  fireEvent.change(inputEmailNode, {target: { value: 'tester@localhost.io' } });
  fireEvent.click(nextButtonNode);
  expect(xhttp.get.mock.calls[0][0]).toMatch(/(\/user\?u=.*&app=.*|\/user\?app=.*&u=.*)/);
  await waitfor( () => expect(container).toMatchSnapshot() );

  const backButtonNode = screen.getAllByText('Back')[0];
  fireEvent.click(backButtonNode);
  await waitfor( () => expect(container).toMatchSnapshot() );

});

