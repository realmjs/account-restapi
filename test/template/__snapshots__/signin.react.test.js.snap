// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot of first scene (email) after mounted 1`] = `
<div
  className="w3-container"
  style={
    Object {
      "maxWidth": "460px",
      "padding": "10px 12px",
    }
  }
>
  <div>
    <div
      className=""
      style={
        Object {
          "display": "block",
        }
      }
    >
      <div>
        <header>
          <span
            className="w3-button w3-right w3-red"
            onClick={[MockFunction]}
            style={
              Object {
                "display": "none",
              }
            }
          >
            ×
          </span>
          <h3
            className="w3-text-blue"
            style={
              Object {
                "fontWeight": "bold",
              }
            }
          >
             Login 
          </h3>
        </header>
        <p
          className="w3-text-blue-grey"
        >
          Please enter your email
        </p>
        <p>
          <label
            className="w3-text-grey"
          >
             Email 
          </label>
          <label
            className="w3-right w3-text-red"
          >
             
            
             
          </label>
          <input
            aria-label="email"
            className="w3-input w3-border "
            disabled={false}
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="email@example.com"
            type="email"
            value=""
          />
        </p>
        <div
          style={
            Object {
              "marginBottom": "42px",
            }
          }
        >
          <div
            className="w3-cell-row"
          >
            <div
              className="w3-cell"
            >
              <label
                className="w3-text-red w3-large"
              >
                <a
                  href="/form?name=signup&app=account"
                >
                   Create New Account 
                </a>
              </label>
            </div>
            <div
              className="w3-cell"
              style={
                Object {
                  "textAlign": "right",
                }
              }
            >
              <button
                className="w3-button w3-blue "
                disabled={false}
                onClick={[Function]}
                type="submit"
              >
                Next 
                 
                <i
                  className="fa fa-chevron-right"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className=""
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div>
        <header>
          <span
            className="w3-button w3-right w3-red"
            onClick={[MockFunction]}
            style={
              Object {
                "display": "none",
              }
            }
          >
            ×
          </span>
          <div>
            <button
              className="w3-button w3-border"
              onClick={[Function]}
              style={
                Object {
                  "marginBottom": "16px",
                }
              }
            >
              <i
                className="fa fa-arrow-left"
              />
               
              Back
            </button>
          </div>
        </header>
        <div
          style={
            Object {
              "display": "block",
            }
          }
        >
          <div
            className="w3-text-blue"
          >
            <h3>
               
              
               
            </h3>
          </div>
          <div>
            <p>
              <label
                className="w3-text-grey"
              >
                Password
              </label>
              <label
                className="w3-right w3-text-red"
              >
                 
                
                 
              </label>
              <input
                className="w3-input w3-border "
                onChange={[Function]}
                onKeyUp={[Function]}
                placeholder="password"
                type="password"
                value=""
              />
            </p>
            <div
              style={
                Object {
                  "marginBottom": "42px",
                }
              }
            >
              <div
                className="w3-cell-row"
              >
                <div
                  className="w3-cell"
                >
                  <label
                    className="w3-text-orange "
                  >
                    <span
                      onClick={[Function]}
                      style={
                        Object {
                          "cursor": "pointer",
                        }
                      }
                    >
                       Forgot your password 
                    </span>
                  </label>
                </div>
                <div
                  className="w3-cell"
                  style={
                    Object {
                      "textAlign": "right",
                    }
                  }
                >
                  <button
                    className="w3-button w3-blue"
                    disabled={false}
                    onClick={[Function]}
                  >
                    Submit 
                     
                    <i
                      className="fa fa-level-down fa-rotate-90"
                      style={
                        Object {
                          "marginLeft": "4px",
                        }
                      }
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          style={
            Object {
              "display": "none",
            }
          }
        />
      </div>
    </div>
    <div
      className="w3-animate-top"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div>
          <h3>
             Welcome 
             
          </h3>
        </div>
        <p>
          You have signed in.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should move to Password scene after entering a registered email and back after clicking back button 1`] = `
<div>
  <div
    class="w3-container"
    style="padding: 10px 12px; max-width: 460px;"
  >
    <div>
      <div
        class=""
        style="display: none;"
      >
        <div>
          <header>
            <span
              class="w3-button w3-right w3-red"
              style="display: none;"
            >
              ×
            </span>
            <h3
              class="w3-text-blue"
              style="font-weight: bold;"
            >
               Login 
            </h3>
          </header>
          <p
            class="w3-text-blue-grey"
          >
            Please enter your email
          </p>
          <p>
            <label
              class="w3-text-grey"
            >
               Email 
            </label>
            <label
              class="w3-right w3-text-red"
            >
               
              
               
            </label>
            <input
              aria-label="email"
              class="w3-input w3-border "
              placeholder="email@example.com"
              type="email"
              value="tester@localhost.io"
            />
          </p>
          <div
            style="margin-bottom: 42px;"
          >
            <div
              class="w3-cell-row"
            >
              <div
                class="w3-cell"
              >
                <label
                  class="w3-text-red w3-large"
                >
                  <a
                    href="/form?name=signup&app=account"
                  >
                     Create New Account 
                  </a>
                </label>
              </div>
              <div
                class="w3-cell"
                style="text-align: right;"
              >
                <button
                  class="w3-button w3-blue "
                  type="submit"
                >
                  Next 
                   
                  <i
                    class="fa fa-chevron-right"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class=""
        style="display: block;"
      >
        <div>
          <header>
            <span
              class="w3-button w3-right w3-red"
              style="display: none;"
            >
              ×
            </span>
            <div>
              <button
                class="w3-button w3-border"
                style="margin-bottom: 16px;"
              >
                <i
                  class="fa fa-arrow-left"
                />
                 
                Back
              </button>
            </div>
          </header>
          <div
            style="display: block;"
          >
            <div
              class="w3-text-blue"
            >
              <h3>
                 
                tester@localhost.io
                 
              </h3>
            </div>
            <div>
              <p>
                <label
                  class="w3-text-grey"
                >
                  Password
                </label>
                <label
                  class="w3-right w3-text-red"
                >
                   
                  
                   
                </label>
                <input
                  class="w3-input w3-border "
                  placeholder="password"
                  type="password"
                  value=""
                />
              </p>
              <div
                style="margin-bottom: 42px;"
              >
                <div
                  class="w3-cell-row"
                >
                  <div
                    class="w3-cell"
                  >
                    <label
                      class="w3-text-orange "
                    >
                      <span
                        style="cursor: pointer;"
                      >
                         Forgot your password 
                      </span>
                    </label>
                  </div>
                  <div
                    class="w3-cell"
                    style="text-align: right;"
                  >
                    <button
                      class="w3-button w3-blue"
                    >
                      Submit 
                       
                      <i
                        class="fa fa-level-down fa-rotate-90"
                        style="margin-left: 4px;"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style="display: none;"
          />
        </div>
      </div>
      <div
        class="w3-animate-top"
        style="display: none;"
      >
        <div
          style="text-align: center;"
        >
          <div>
            <h3>
               Welcome 
               
            </h3>
          </div>
          <p>
            You have signed in.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should move to Password scene after entering a registered email and back after clicking back button 2`] = `
<div>
  <div
    class="w3-container"
    style="padding: 10px 12px; max-width: 460px;"
  >
    <div>
      <div
        class=""
        style="display: block;"
      >
        <div>
          <header>
            <span
              class="w3-button w3-right w3-red"
              style="display: none;"
            >
              ×
            </span>
            <h3
              class="w3-text-blue"
              style="font-weight: bold;"
            >
               Login 
            </h3>
          </header>
          <p
            class="w3-text-blue-grey"
          >
            Please enter your email
          </p>
          <p>
            <label
              class="w3-text-grey"
            >
               Email 
            </label>
            <label
              class="w3-right w3-text-red"
            >
               
              
               
            </label>
            <input
              aria-label="email"
              class="w3-input w3-border "
              placeholder="email@example.com"
              type="email"
              value="tester@localhost.io"
            />
          </p>
          <div
            style="margin-bottom: 42px;"
          >
            <div
              class="w3-cell-row"
            >
              <div
                class="w3-cell"
              >
                <label
                  class="w3-text-red w3-large"
                >
                  <a
                    href="/form?name=signup&app=account"
                  >
                     Create New Account 
                  </a>
                </label>
              </div>
              <div
                class="w3-cell"
                style="text-align: right;"
              >
                <button
                  class="w3-button w3-blue "
                  type="submit"
                >
                  Next 
                   
                  <i
                    class="fa fa-chevron-right"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class=""
        style="display: none;"
      >
        <div>
          <header>
            <span
              class="w3-button w3-right w3-red"
              style="display: none;"
            >
              ×
            </span>
            <div>
              <button
                class="w3-button w3-border"
                style="margin-bottom: 16px;"
              >
                <i
                  class="fa fa-arrow-left"
                />
                 
                Back
              </button>
            </div>
          </header>
          <div
            style="display: block;"
          >
            <div
              class="w3-text-blue"
            >
              <h3>
                 
                tester@localhost.io
                 
              </h3>
            </div>
            <div>
              <p>
                <label
                  class="w3-text-grey"
                >
                  Password
                </label>
                <label
                  class="w3-right w3-text-red"
                >
                   
                  
                   
                </label>
                <input
                  class="w3-input w3-border "
                  placeholder="password"
                  type="password"
                  value=""
                />
              </p>
              <div
                style="margin-bottom: 42px;"
              >
                <div
                  class="w3-cell-row"
                >
                  <div
                    class="w3-cell"
                  >
                    <label
                      class="w3-text-orange "
                    >
                      <span
                        style="cursor: pointer;"
                      >
                         Forgot your password 
                      </span>
                    </label>
                  </div>
                  <div
                    class="w3-cell"
                    style="text-align: right;"
                  >
                    <button
                      class="w3-button w3-blue"
                    >
                      Submit 
                       
                      <i
                        class="fa fa-level-down fa-rotate-90"
                        style="margin-left: 4px;"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style="display: none;"
          />
        </div>
      </div>
      <div
        class="w3-animate-top"
        style="display: none;"
      >
        <div
          style="text-align: center;"
        >
          <div>
            <h3>
               Welcome 
               
            </h3>
          </div>
          <p>
            You have signed in.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
